main.go

```diff
@@ -7,103 +7,151 @@ import (
+	latestBlockHash, err := client.GetBlockHash(blockCount)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		// leftBlockHash, err := client.GetBlockHash(leftBlockHeight)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	latestBlock, err := client.GetBlockVerbose(latestBlockHash)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		// leftBlock, err := client.GetBlockVerbose(leftBlockHash)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	latestBlockTime := latestBlock.Time
 
-		// rightBlockHash, err := client.GetBlockHash(rightBlockHeight)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	// Check if the target time is in the future
+	if targetTime > latestBlockTime {
+		timeDifference := targetTime - latestBlockTime
+		estimatedFutureBlocks := timeDifference / averageBlockTime
+		return blockCount + estimatedFutureBlocks
+	}
 
-		// rightBlock, err := client.GetBlockVerbose(rightBlockHash)
-		// 	if err != nil {
-		// 		log.Fatal(err)
-		// 	}
+	// Perform binary search for past blocks
+	var leftBlockHeight, rightBlockHeight int64 = 0, blockCount
+
+	for leftBlockHeight <= rightBlockHeight {
+		midBlockHeight := (leftBlockHeight + rightBlockHeight) / 2
 
-		midBlockTime := midBlock.Time
-		// leftBlockTime := leftBlock.Time
-		// rightBlockTime := rightBlock.Time
+		midBlockTime, err := getBlockTime(midBlockHeight)
+		if err != nil {
+			log.Printf("Error getting block time: %v", err)
+			return -1
+		}
 
 		if midBlockTime == targetTime {
+			cacheMutex.Lock()
+			cacheMap[targetTime] = cacheEntry{height: midBlockHeight, timestamp: time.Now()}
+			cacheMutex.Unlock()
 			return midBlockHeight
 		} else if midBlockTime < targetTime {
 			leftBlockHeight = midBlockHeight + 1
 		} else {
 			rightBlockHeight = midBlockHeight - 1
 		}
 	}
-	return (leftBlockHeight)
+
+	result := leftBlockHeight
+	cacheMutex.Lock()
+	cacheMap[targetTime] = cacheEntry{height: result, timestamp: time.Now()}
+	cacheMutex.Unlock()
+	return result
 }
 
 func main() {
-	// blockchain stuff
-	err := godotenv.Load(".env")
-	if err != nil {
-		log.Fatalf("Error loading .env file")
-	}
-	username := os.Getenv("BTCUSER")
-	password := os.Getenv("PASSWORD")
-	host := os.Getenv("HOST")
-	// Connect to local bitcoin core RPC server using HTTP POST mode.
-	connCfg := &rpcclient.ConnConfig{
-		// Host:         "localhost:8332",
-		Host:         host,
-		User:         username,
-		Pass:         password,
-		HTTPPostMode: true, // Bitcoin core only supports HTTP POST mode
-		DisableTLS:   true, // Bitcoin core does not provide TLS by default
-	}
-	// Notice the notification parameter is nil since notifications are
-	// not supported in HTTP POST mode.
-	client, err := rpcclient.New(connCfg, nil)
-	if err != nil {
-		log.Fatal(err)
-	}
 	defer client.Shutdown()
 
-	// // Get the current block count.
-	// blockCount, err := client.GetBlockCount()
-	// if err != nil {
-	// 	log.Fatal(err)
-	// }
+	// Start cache cleanup goroutine
+	go clearCachePeriodically()
```

Comment:
What is the purpose of using a goroutine?

Reply:
It spins up the process to clear the cache in a separate thread instead of blocking the main thread always.